### DISCUSSION TITLE
How to configure YOLO keypoints to work with a common key point for multiple body parts?

### DISCUSSION THREAD

**Discussion by Hezonar** [ultralytics/ultralytics#17820] [Q&A]
Hello!

I'm training a YOLOv8 model for the keypoint problem and working with animal leg markings. The following questions arose:

1. Will the model correctly distinguish between key points, such as the left and right leg, if they are specified separately? How does she assess which leg is detected?
2. Is it possible to specify a common point “leg” so as not to divide it into left and right? For example, use one keypoint (one category) for all four legs?

I have attached a picture to make it more clear. 
If the dog goes to the left, then the leg closest to us will be the left. But if she goes to the right, then the closest leg will be the right one.
![image](https://github.com/user-attachments/assets/a9b48b91-b062-4b3a-8c76-cdd703132407)
![image](https://github.com/user-attachments/assets/5e0bc5aa-1b1f-4bbb-be65-a2d2ebddb366)

In my task, determining the position of the legs is quite important. The pictures of the legs are not that good, so I would like to understand better. Can YOLO confuse back/front legs and left/right? How does the classification and definition of a point occur within the YOLO model itself?

Thank you for your help! I would appreciate any recommendations or clarifications regarding keypoints.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Hezonar, thank you for your interest in Ultralytics 🚀 and for providing detailed context about your keypoint detection task with YOLOv8! Your project involving animal leg markings sounds fascinating 🐾.

We recommend reviewing the [Docs](https://docs.ultralytics.com), specifically the [Keypoints Guide](https://docs.ultralytics.com/tasks/keypoints/) for insights on how keypoint detection is implemented in YOLOv8, which may address some of your questions.

For your queries:

1. If this concerns a potential 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) along with any specifics such as dataset details or model configurations. This will help us debug it effectively.
2. If this is a ❓ customization or general question, additional information like dataset details, training logs, and examples of input images would be greatly helpful to provide further guidance.

Also, please ensure you are using the latest `ultralytics` package. You can upgrade with:
```bash
pip install -U ultralytics
```

If you'd like to experiment with running YOLOv8 in various environments, here are some recommended options that come preconfigured with dependencies:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For community-driven support and guidance, consider joining the Ultralytics community where it suits you best:
- [Discord](https://ultralytics.com/discord) for real-time conversations 🎧
- [Discourse](https://community.ultralytics.com) for in-depth technical discussions
- [Reddit](https://reddit.com/r/ultralytics) for sharing knowledge and experiences 💡

Finally, ensure your YOLOv8 environment matches the Python>=3.8 and PyTorch>=1.8 requirements listed in the [Ultralytics requirements file](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

This is an automated response 😊, but rest assured an Ultralytics engineer will assist you directly as soon as possible!

**Comment by glenn-jocher** [comment]
Thank you for sharing your question. Here's how YOLOv8 handles keypoints and how you can configure them:

1. Yes, YOLOv8 can distinguish between separate keypoints like left and right legs, as long as your training data has consistent annotations for these keypoints and their corresponding labels. The model evaluates each keypoint's position relative to the bounding box and other keypoints in the image during inference.

2. It is possible to define a single keypoint category (e.g., "leg") for all legs instead of separate ones (e.g., left/right or front/back). However, this will limit the model's ability to differentiate between specific legs. To implement this, ensure your dataset annotations reflect this aggregated keypoint strategy.

For your scenario, consistent and precise annotation is crucial. If maintaining the distinction between leg types is important, consider training the model with clearer labels and more diverse images to reduce confusion. You can learn more about pose estimation in YOLO models [here](https://docs.ultralytics.com/tasks/pose/).

**Comment by Y-T-G** [comment]
Each leg will require a distinct keypoint. And you also need to specify `flip_idx` in the yaml file to say they're symmetrical.

https://github.com/ultralytics/ultralytics/issues/17736
